x-pack/plugins/security/server/lib/watch_status_and_license_to_initialize.test.js

```diff
@@ -0,0 +1,221 @@
+
+  watchStatusAndLicenseToInitialize(mockXpackMainPlugin, downstreamPlugin, initializeMock);
+
+  expect(initializeMock).toHaveBeenCalledTimes(1);
+  expect(initializeMock).toHaveBeenCalledWith(licenseCheckResults);
+  downstreamPlugin.status.red.mockImplementation(message => {
+    expect(message).toBe(errorMessage);
+    done();
+  });
+});
+
+test(`calls initialize twice when it gets a new license and the status is green`, (done) => {
+  const pluginId = 'foo-plugin';
+  const { mockXpackMainPlugin, mockFeature } = createMockXpackMainPluginAndFeature(pluginId);
+  mockXpackMainPlugin.mock.setStatus('green');
+  const firstLicenseCheckResults = Symbol();
+  const secondLicenseCheckResults = Symbol();
+  mockFeature.mock.setLicenseCheckResults(firstLicenseCheckResults);
+  const downstreamPlugin = createMockDownstreamPlugin(pluginId);
+  const initializeMock = jest.fn().mockImplementation(() => Promise.resolve());
+
+  let count = 0;
+  downstreamPlugin.status.green.mockImplementation(message => {
+    expect(message).toBe('Ready');
+    ++count;
+    if (count === 1) {
+      mockFeature.mock.setLicenseCheckResults(secondLicenseCheckResults);
+      mockFeature.mock.triggerLicenseChange();
+    }
+    if (count === 2) {
+      expect(initializeMock).toHaveBeenCalledWith(firstLicenseCheckResults);
+      expect(initializeMock).toHaveBeenCalledWith(secondLicenseCheckResults);
+      expect(initializeMock).toHaveBeenCalledTimes(2);
+      done();
+    }
+  });
+
+  watchStatusAndLicenseToInitialize(mockXpackMainPlugin, downstreamPlugin, initializeMock);
+});
+
+test(`doesn't call initialize twice when it gets a new license when the status isn't green`, (done) => {
+  const pluginId = 'foo-plugin';
+  const redMessage = 'the red message';
+  const { mockXpackMainPlugin, mockFeature } = createMockXpackMainPluginAndFeature(pluginId);
+  mockXpackMainPlugin.mock.setStatus('green');
+  const firstLicenseCheckResults = Symbol();
+  const secondLicenseCheckResults = Symbol();
+  mockFeature.mock.setLicenseCheckResults(firstLicenseCheckResults);
+  const downstreamPlugin = createMockDownstreamPlugin(pluginId);
+  const initializeMock = jest.fn().mockImplementation(() => Promise.resolve());
+
+  downstreamPlugin.status.green.mockImplementation(message => {
+    expect(message).toBe('Ready');
+    mockXpackMainPlugin.mock.setStatus('red', redMessage);
+    mockFeature.mock.setLicenseCheckResults(secondLicenseCheckResults);
+    mockFeature.mock.triggerLicenseChange();
+  });
+
+  downstreamPlugin.status.red.mockImplementation(message => {
+    expect(message).toBe(redMessage);
+    expect(initializeMock).toHaveBeenCalledTimes(1);
+    expect(initializeMock).toHaveBeenCalledWith(firstLicenseCheckResults);
+    done();
+  });
+
+  watchStatusAndLicenseToInitialize(mockXpackMainPlugin, downstreamPlugin, initializeMock);
+});
+
+test(`calls initialize twice when the status changes to green twice`, (done) => {
```

Comment:
What is the use case for reinitializing when transitioning from green -> green?

Reply:
In real-world scenarios, we'd actually go green -> red -> green (or some other state in the middle), I skipped a step in the test to simplify it, I mainly wanted to test that initialize was called whenever it goes green.